### PR TITLE
Fix conversions from subnormal System.Half values

### LINKED ISSUE #46515: Half to Double/Float conversion is broken for subnormal values
### Description

`Half` numbers with binary values between 0x0200 and 0x03FF are converted to approximately half their value when cast to `Single` and `Double`. 0x03FF corresponds to the largest positive subnormal number.

```c#
// largest subnormal, binary 0x03ff
Half h1 = (Half)0.000060975552;
Console.WriteLine("0x{0:X4}", h1.GetHashCode());
// Prints 3.0487776E-05, which is off by a factor of two
Console.WriteLine("{0:R}", (Single)h1);

// binary 0x0200
Half h2 = (Half)3.05E-05;
Console.WriteLine("0x{0:X4}", h2.GetHashCode());
// Prints 1.5258789E-05, which is off by a factor of two
Console.WriteLine("{0:R}", (Single)h2);

// binary 0x01FF
Half h3 = (Half)3.046E-05;
Console.WriteLine("0x{0:X4}", h3.GetHashCode());
// Prints 3.0457973E-05, which seems about right
Console.WriteLine("{0:R}", (Single)h3);

// binary 0x0400
Half h4 = (Half)6.104E-05;
Console.WriteLine("0x{0:X4}", h4.GetHashCode());
// Prints 6.1035156E-05, which seems about right
Console.WriteLine("{0:R}", (Single)h4);
```

### Configuration

.NET version: 5.0.101
OS: Linux 5.4.0-45-generic
Distro: Linux Mint 20
Architecture: x86_64

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/46515.

CC. @pgovind, @ecorm

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Half.cs [hunk 1]
```diff
@@ -683,10 +683,10 @@ private static double CreateDoubleNaN(bool sign, ulong significand)
         }
 
         private static float CreateSingle(bool sign, byte exp, uint sig)
-            => BitConverter.Int32BitsToSingle((int)(((sign ? 1U : 0U) << float.SignShift) | ((uint)exp << float.ExponentShift) | sig));
+            => BitConverter.Int32BitsToSingle((int)(((sign ? 1U : 0U) << float.SignShift) + ((uint)exp << float.ExponentShift) + sig));
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Half.cs:686]: The issue was that this was doing a bitwise or when it should have been using addition.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Half.cs:686]: The original implementation was doing the correct thing: https://github.com/ucb-bar/berkeley-softfloat-3/blob/master/source/include/internals.h#L107

**Review by pgovind — APPROVED** [review]
Ah yup, my bad. LGTM

**Review by jeffhandley — APPROVED** [review]
We should submit this for 5.0 servicing approval. In the servicing template on the backport PR, please include notes for which boundary conditions we were previously testing and which ones we missed.

We will need to backport the test updates too (first), which can be done without approval.

**Comment by ecorm** [comment]
I've just uncovered another corner case where .NET's `Half` does not agree with http://half.sourceforge.net/ . When converting from Single(MinSubnormal/2) to `Half` (where MinSubnormal/2 is 0x33000000), the Sourceforge version makes it underflow to zero, whereas .NET makes it round up to `Half.Epsilon`. I'm not sure which is correct. I'm currently checking if @tannergooding 's fix makes both implementations agree.

**Comment by ecorm** [comment]
> Ah yup, my bad. LGTM

I made the same mistake when I adapted http://half.sourceforge.net/ for my own use. I thought the `+` was an attempt to optimize a bitwise OR, but it turns out it really needs to perform addition for the rounding logic to work.

**Comment by ecorm** [comment]
> I'm currently checking if @tannergooding 's fix makes both implementations agree.

Nope, they still disagree for the (Half)(`BitConverter.Int32BitsToSingle(0x33000000)`) case. Perhaps this should be a separate issue.

**Comment by tannergooding** [comment]
I've found the issue and a local fix, will push it up with a commit momentarily

File: src/libraries/System.Private.CoreLib/src/System/Half.cs [hunk 2]
```diff
@@ -635,6 +635,7 @@ private static ushort RoundPackToHalf(bool sign, short exp, ushort sig)
                 {
                     sig = (ushort)ShiftRightJam(sig, -exp);
                     exp = 0;
+                    roundBits = sig & 0xF;
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Half.cs:638]: The other issue was that this was missing adjusting the roundBits based on the updated sig.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Half.cs:638]: The original implementation was doing the correct thing: https://github.com/ucb-bar/berkeley-softfloat-3/blob/b64af41c3276f97f0e181920400ee056b9c88037/source/s_roundPackToF16.c#L77

I imagine this was accidentally dropped when eliding the unnecessary exception handling logic.

**Comment by ecorm** [comment]
> I'm not sure which is correct.

I think the Sourceforge implementation was correct for the MinSubnormal/2 case. Ties round to even, so `UInt16BitsToHalf(0)` being even would be the correct choice, whereas `UInt16BitsToHalf(0x0001)` would be odd.

Sorry if this all sounds elementary, but I've never needed deal with the gory inner workings of IEEE floats before. :smile:

**Comment by ecorm** [comment]
Feel free to borrow (and verify) some of my test vectors from here while you're at it: https://gist.github.com/ecorm/3b5963749752f2cc2f96fb493d29fcba

**Pull request merged**